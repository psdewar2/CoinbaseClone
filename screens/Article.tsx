import {useNavigation, useRoute} from '@react-navigation/core';
import {useTheme} from '@react-navigation/native';
import React, {useEffect, useRef, useState} from 'react';
import {Text, View} from 'react-native';
import {TouchableOpacity} from 'react-native-gesture-handler';
import {SafeAreaView} from 'react-native-safe-area-context';
import {WebView} from 'react-native-webview';
import {COINBASE_BLUE} from '../Constants';

export function Article() {
  const {colors} = useTheme();
  const {goBack} = useNavigation();
  const {
    params: {uri},
  } = useRoute() as any;
  const [progress, setProgress] = useState(0);
  const webviewRef = useRef<WebView>(null);

  function extractHostname(url: string) {
    let hostname;
    // find & remove protocol (http, ftp, etc.) and get hostname

    if (url.indexOf('//') > -1) {
      hostname = url.split('/')[2];
    } else {
      hostname = url.split('/')[0];
    }
    // find & remove port number
    hostname = hostname.split(':')[0];
    // find & remove "?"
    hostname = hostname.split('?')[0];

    return hostname;
  }

  useEffect(() => {
    let timeout: NodeJS.Timeout;
    if (progress === 1) {
      timeout = setTimeout(() => setProgress(0), 100);
    }

    return () => {
      timeout && clearTimeout(timeout);
    };
  }, [progress]);

  return (
    <SafeAreaView style={{flex: 1, justifyContent: 'space-between'}}>
      <View
        style={{
          flexDirection: 'row',
          justifyContent: 'space-between',
          marginTop: 6,
          marginHorizontal: 14,
          marginBottom: 14,
        }}>
        <TouchableOpacity onPress={goBack}>
          <Text style={{fontSize: 16, fontWeight: '600', color: COINBASE_BLUE}}>
            Close
          </Text>
        </TouchableOpacity>
        <Text
          numberOfLines={1}
          style={{
            color: colors.text,
            flex: 1,
            fontSize: 16,
            fontWeight: '600',
            textAlign: 'center',
            paddingHorizontal: 10,
          }}>
          {extractHostname(uri)}
        </Text>
        <TouchableOpacity onPress={() => webviewRef.current?.reload()}>
          <Text style={{fontSize: 16, fontWeight: '600', color: COINBASE_BLUE}}>
            Reload
          </Text>
        </TouchableOpacity>
      </View>
      <View
        style={{
          opacity: progress ? 1 : 0,
          borderColor: COINBASE_BLUE,
          borderWidth: 2,
          width: `${progress * 100}%`,
        }}
      />
      <WebView
        ref={webviewRef}
        source={{uri}}
        onLoadProgress={event => setProgress(event.nativeEvent.progress)}
      />
    </SafeAreaView>
  );
}
