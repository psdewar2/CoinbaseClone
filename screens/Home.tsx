import {useNavigation} from '@react-navigation/core';
import {useTheme} from '@react-navigation/native';
import React from 'react';
import {Image, ScrollView, Text, TouchableOpacity, View} from 'react-native';
import {SafeAreaView} from 'react-native-safe-area-context';
import {News} from '../components/News';
import {TopMovers} from '../components/TopMovers';
import {Watchlist} from '../components/Watchlist';

export function Home() {
  const {colors, dark} = useTheme();
  const {navigate} = useNavigation() as any; // not best practice
  return (
    <SafeAreaView
      edges={['top']}
      style={{flex: 1, backgroundColor: dark ? '#111' : '#FFF'}}>
      <ScrollView>
        <View
          style={{
            alignItems: 'center',
            justifyContent: 'center',
            paddingTop: 50,
          }}>
          <Image
            source={require('../assets/wallet.jpeg')}
            style={{width: 220, height: 220}}
          />
          <Text
            style={{
              color: colors.text,
              fontSize: 20,
              fontWeight: '600',
              paddingTop: 10,
            }}>
            Welcome to coinbase
          </Text>
          <Text
            style={{
              fontSize: 16,
              fontWeight: '400',
              color: '#5d616d',
              paddingTop: 10,
            }}>
            Make your first investment today
          </Text>
          <View style={{paddingTop: 30}}>
            <TouchableOpacity
              disabled
              style={{
                backgroundColor: '#2150f5',
                borderRadius: 8,
                paddingVertical: 17,
                paddingHorizontal: 100,
              }}>
              <Text
                style={{
                  fontSize: 16,
                  color: 'white',
                  fontWeight: '600',
                  alignSelf: 'center',
                }}>
                Fund Account
              </Text>
            </TouchableOpacity>
          </View>
        </View>
        <Watchlist />
        <TopMovers />
        <News onPress={(uri: string) => navigate('Article', {uri})} />
      </ScrollView>
    </SafeAreaView>
  );
}
