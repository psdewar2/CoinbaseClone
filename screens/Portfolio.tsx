import {useTheme} from '@react-navigation/native';
import React, {useEffect, useState} from 'react';
import {ActivityIndicator, Image, ScrollView, Text, View} from 'react-native';
import {SafeAreaView} from 'react-native-safe-area-context';

export function Portfolio() {
  const {colors, dark} = useTheme();
  const [isLoading, setIsLoading] = useState(false);
  const [data, setData] = useState<
    {id: string; image: {large: string}; name: string; symbol: string}[]
  >([]);
  useEffect(() => {
    async function getCoins() {
      setIsLoading(true);
      try {
        const response = await fetch('https://api.coingecko.com/api/v3/coins/');
        setData(await response.json());
      } catch (e) {
      } finally {
        setIsLoading(false);
      }
    }
    getCoins();
  }, []);

  return (
    <SafeAreaView
      edges={['top']}
      style={{flex: 1, backgroundColor: dark ? '#111' : '#FFF'}}>
      {isLoading ? (
        <View
          style={{
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
          }}>
          <ActivityIndicator size="large" />
        </View>
      ) : (
        <ScrollView>
          <View style={{paddingTop: 50, paddingHorizontal: 20}}>
            <Text style={{color: colors.text, fontSize: 14, fontWeight: '500'}}>
              Portfolio Balance
            </Text>
            <Text
              style={{
                color: colors.text,
                fontSize: 29,
                fontWeight: 'bold',
                paddingTop: 5,
              }}>
              $0.00
            </Text>
            <View style={{paddingBottom: 20}}>
              {data.map(({id, image, name, symbol}) => (
                <View key={id}>
                  <View
                    style={{
                      paddingTop: 25,
                      flexDirection: 'row',
                      justifyContent: 'space-between',
                      alignItems: 'center',
                    }}>
                    <View>
                      <Image
                        source={{uri: image.large}}
                        style={{
                          width: 32,
                          height: 32,
                          borderRadius: 16,
                          borderWidth: 0.5,
                          borderColor: colors.border,
                        }}
                      />
                    </View>
                    <View style={{flex: 1, paddingLeft: 15}}>
                      <Text
                        style={{
                          fontSize: 17,
                          fontWeight: '400',
                          color: colors.text,
                        }}>
                        {name}
                      </Text>
                    </View>
                    <View style={{alignItems: 'flex-end'}}>
                      <Text
                        style={{
                          fontSize: 16,
                          fontWeight: '300',
                          color: colors.text,
                        }}>
                        $0.00
                      </Text>
                      <Text
                        style={{
                          fontSize: 14,
                          fontWeight: '300',
                          color: '#5d616d',
                        }}>
                        0 {symbol}
                      </Text>
                    </View>
                  </View>
                </View>
              ))}
            </View>
          </View>
        </ScrollView>
      )}
    </SafeAreaView>
  );
}
