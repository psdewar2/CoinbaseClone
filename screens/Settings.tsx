import {useTheme} from '@react-navigation/native';
import React from 'react';
import {Text} from 'react-native';
import {SafeAreaView} from 'react-native-safe-area-context';

export function Settings() {
  const {colors} = useTheme();
  return (
    <SafeAreaView>
      <Text style={{color: colors.text}}>For You</Text>
    </SafeAreaView>
  );
}
