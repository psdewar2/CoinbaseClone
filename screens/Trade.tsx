import {useTheme} from '@react-navigation/native';
import React from 'react';
import {Text} from 'react-native';
import {SafeAreaView} from 'react-native-safe-area-context';

export function Trade() {
  const {colors} = useTheme();
  return (
    <SafeAreaView>
      <Text style={{color: colors.text}}>Trade</Text>
    </SafeAreaView>
  );
}
