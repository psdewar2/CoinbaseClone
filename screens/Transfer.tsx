import {useNavigation} from '@react-navigation/core';
import {useTheme} from '@react-navigation/native';
import React from 'react';
import {Text, View} from 'react-native';
import {TouchableOpacity} from 'react-native-gesture-handler';

const actions = [
  {
    name: 'Buy',
    desc: 'Buy crypto with cash',
    icon: '+',
  },
  {
    name: 'Sell',
    desc: 'Sell crypto for cash',
    icon: '-',
  },
  {
    name: 'Convert',
    desc: 'Convert one crypto to another',
    icon: 'C',
  },
  {
    name: 'Send',
    desc: 'Send crypto to another wallet',
    icon: 'S',
  },
  {
    name: 'Receive',
    desc: 'Receive crypto from another wallet',
    icon: 'R',
  },
  {
    name: 'Add cash / Cash out',
    desc: 'Transfer balance to or from your bank',
    icon: 'B',
  },
];

export function Transfer() {
  const {colors, dark} = useTheme();
  const {goBack} = useNavigation();
  return (
    <View style={{flex: 1, backgroundColor: dark ? '#111' : '#FFF'}}>
      <TouchableOpacity
        onPress={goBack}
        style={{paddingLeft: 20, paddingTop: 20}}>
        <Text style={{color: colors.text, fontWeight: '600'}}>Close</Text>
      </TouchableOpacity>
      {actions.map(({name, desc, icon}) => {
        return (
          <TouchableOpacity
            key={name}
            style={{flexDirection: 'row', padding: 20}}>
            <View
              style={{
                justifyContent: 'center',
                alignItems: 'center',
                width: 60,
              }}>
              <Text
                style={{fontSize: 20, fontWeight: '600', color: colors.text}}>
                {icon}
              </Text>
            </View>
            <View>
              <Text
                style={{
                  fontWeight: '600',
                  marginBottom: 4,
                  color: colors.text,
                }}>
                {name}
              </Text>
              <Text style={{color: 'gray'}}>{desc}</Text>
            </View>
          </TouchableOpacity>
        );
      })}
    </View>
  );
}
