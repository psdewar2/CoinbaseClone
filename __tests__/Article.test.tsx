import React from 'react';
import {render} from '@testing-library/react-native';
import {Article} from '../screens/Article';
import {NavigationContainer} from '@react-navigation/native';

describe('Article', () => {
  // Arrange
  const component = (
    <NavigationContainer>
      <Article />
    </NavigationContainer>
  );

  it('renders without crashing', async () => {
    // Act
    const rendererJson = render(component).toJSON();
    // Assert
    expect(rendererJson).toMatchSnapshot();
  });
});
