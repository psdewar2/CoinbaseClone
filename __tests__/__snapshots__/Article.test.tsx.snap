// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article renders without crashing 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 14,
        "marginHorizontal": 14,
        "marginTop": 6,
      }
    }
  >
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
    >
      <View
        accessible={true}
        collapsable={false}
        nativeID="animatedComponent"
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4169e1",
              "fontSize": 16,
              "fontWeight": "600",
            }
          }
        >
          Close
        </Text>
      </View>
    </RNGestureHandlerButton>
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "rgb(28, 28, 30)",
          "flex": 1,
          "fontSize": 16,
          "fontWeight": "600",
          "paddingHorizontal": 10,
          "textAlign": "center",
        }
      }
    >
      coinbase.com
    </Text>
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
    >
      <View
        accessible={true}
        collapsable={false}
        nativeID="animatedComponent"
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4169e1",
              "fontSize": 16,
              "fontWeight": "600",
            }
          }
        >
          Reload
        </Text>
      </View>
    </RNGestureHandlerButton>
  </View>
  <View
    style={
      Object {
        "borderColor": "#4169e1",
        "borderWidth": 2,
        "opacity": 0,
        "width": "0%",
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <RNCWebView
      cacheEnabled={true}
      injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
      injectedJavaScriptForMainFrameOnly={true}
      javaScriptEnabled={true}
      messagingEnabled={false}
      onContentProcessDidTerminate={[Function]}
      onHttpError={[Function]}
      onLoadProgress={[Function]}
      onLoadingError={[Function]}
      onLoadingFinish={[Function]}
      onLoadingProgress={[Function]}
      onLoadingStart={[Function]}
      onMessage={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      source={
        Object {
          "uri": "https://coinbase.com",
        }
      }
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#ffffff",
          },
          undefined,
        ]
      }
      useSharedProcessPool={true}
    />
  </View>
</RNCSafeAreaView>
`;
